### PR TITLE
MM-28249  Auto follow threads

### PR DESCRIPTION
**Pull request by reflog**
#### Ticket Link

https://mattermost.atlassian.net/browse/MM-28249

    ```release-note
	Added support for thread auto follow
    ```

### CODE REVIEW

File: store/sqlstore/thread_store.go [hunk 1]
```diff
@@ -106,3 +114,69 @@ func (s *SqlThreadStore) Delete(threadId string) error {
...
+
+		if !errors.As(err, &nfErr) {
+			return errors.Wrap(err, "failed to get thread membership")
+		}
+		_, err = s.SaveMembership(&model.ThreadMembership{
+			PostId:      postId,
+			UserId:      userId,
+			Following:   true,
+			LastViewed:  0,
+			LastUpdated: 0,
```
Comment (jespino) [inline store/sqlstore/thread_store.go:176]: The LastUpdate shouldn't be now?

File: store/sqlstore/thread_store.go [hunk 2]
```diff
@@ -106,3 +114,69 @@ func (s *SqlThreadStore) Delete(threadId string) error {
...
+
+func (s *SqlThreadStore) DeleteMembershipForUser(userId string, postId string) error {
+	if _, err := s.GetMaster().Exec("DELETE FROM ThreadMemberships Where PostId = :PostId AND UserId = :UserId", map[string]interface{}{"PostId": postId, "UserId": userId}); err != nil {
+		return errors.Wrap(err, "failed to update thread membership")
+	}
+
+	return nil
+}
+
+func (s *SqlThreadStore) UpdateMembershipFromMention(userId, postId string) error {
```
Comment (jespino) [inline store/sqlstore/thread_store.go:163]: This naming is leaking bussiness logic into the store layer, maybe something like `CreateMembershipIfNeeded`

File: store/sqlstore/thread_store.go [hunk 3]
```diff
@@ -106,3 +114,69 @@ func (s *SqlThreadStore) Delete(threadId string) error {
...
+		_, err = s.SaveMembership(&model.ThreadMembership{
+			PostId:      postId,
+			UserId:      userId,
+			Following:   true,
+			LastViewed:  0,
+			LastUpdated: 0,
+		})
+		return err
+	}
+	_, err = s.UpdateMembership(membership)
```
Comment (jespino) [inline store/sqlstore/thread_store.go:180]: If there is not error getting the membership we are storing again the same membership, I think makes no sense unless you process it, actually. I think if we are leveraging LastUpdated to have Sync api endpoint, probably we want to update the membership if the following is false, to true, and the LastUpdated if that happens.

File: store/sqlstore/thread_store.go [hunk 4]
```diff
@@ -106,3 +116,73 @@ func (s *SqlThreadStore) Delete(threadId string) error {
...
+	}
+
+	return nil
+}
+
+func (s *SqlThreadStore) CreateMembershipIfNeeded(userId, postId string) error {
+	membership, err := s.GetMembershipForUser(userId, postId)
+	now := utils.MillisFromTime(time.Now())
+	if err == nil {
+		membership.Following = true
```
Comment (jespino) [inline store/sqlstore/thread_store.go:169]: Should we check if the membership is already in `Following == true` before we update it? With the current approach we are trying to create the membership or updating it, and changing the update data, even when is not needed. So... the LastUpdated date is going to change, and is going to generate websocket events, extra load in the future "sync" method for mobile... and extra unneeded writes.

File: app/notification.go [hunk 5]
```diff
@@ -165,14 +165,20 @@ func (a *App) SendNotifications(post *model.Post, team *model.Team, channel *mod
...
 			nErr := a.Srv().Store.Channel().IncrementMentionCount(post.ChannelId, userId)
 			if nErr != nil {
 				umc <- model.NewAppError("SendNotifications", "app.channel.increment_mention_count.app_error", nil, nErr.Error(), http.StatusInternalServerError)
-			} else {
-				umc <- nil
+				return
 			}
-
-			close(umc)
+			if *a.Config().ExperimentalSettings.ThreadAutoFollow && post.RootId != "" {
```
Comment (jespino) [inline app/notification.go:174]: I don't think this code belongs here. Looks like you are hijacking the async task to increment the mention count, to do other things. This method is already complex and hard to understand. I would prefer to separate it in its own go routing, and actually encapsulate that in a method to not contribute more to the growth of this method.

File: app/notification.go [hunk 6]
```diff
@@ -165,14 +165,20 @@ func (a *App) SendNotifications(post *model.Post, team *model.Team, channel *mod
...
+				nErr = a.Srv().Store.Thread().CreateMembershipIfNeeded(userId, post.RootId)
+				if nErr != nil {
+					umc <- model.NewAppError("SendNotifications", "app.channel.increment_mention_count.app_error", nil, nErr.Error(), http.StatusInternalServerError)
```
Comment (jespino) [inline app/notification.go:177]: The error message here is reused, and should be its own error message.

**Review by jespino — CHANGES_REQUESTED** [review]
Some small changes needed, but looking pretty good.

File: app/notification.go (hunk 5 above)
Comment (ashishbhate) [reply app/notification.go:174]: Do you mean a separate go routine in this loop? The code is creating a thread membership for all people mentioned in a thread. It seems appropriate to put it in this loop.

Comment (reflog) [reply app/notification.go:174]: @jespino - I also think that the placement for this code is not that bad. This method does similar things, I'm just adding one more 'task'

File: app/notification.go (hunk 6 above)
Comment (reflog) [reply app/notification.go:177]: Oops. will fix

File: store/sqlstore/thread_store.go (hunk 4 above)
Comment (reflog) [reply store/sqlstore/thread_store.go:169]: Right. Will fix

File: app/notification.go (hunk 5 above)
Comment (jespino) [reply app/notification.go:174]: Yes, inside the loop, or later on looping through the `mentionedUsersList`. I don't know what is the best approach there. Probably makes no sense to fail the "notification" if the thread doesn't get followed anyway, so... maybe a different goroutine that does the looping and the execution.

Comment (jespino) [reply app/notification.go]: I disagree with that, the "updateMentionChans" looks like for update mentions count, not for threads autofollow, at least I would have a different name.

At the same time we are binding the success/failure of the incrementnotificationcount to the success or failure of the CreateMembershipIfNeeded, I don't know how much do we want that.

If you feel strong about it, I'm not against it, but looks like there are some rough edges there.

**Review by ashishbhate — APPROVED** [review]

File: app/notification.go [hunk 7]
```diff
@@ -160,19 +160,35 @@ func (a *App) SendNotifications(post *model.Post, team *model.Team, channel *mod
...
+			if *a.Config().ExperimentalSettings.ThreadAutoFollow && post.RootId != "" {
+				nErr := a.Srv().Store.Thread().CreateMembershipIfNeeded(userId, post.RootId)
+				if nErr != nil {
+					umc <- model.NewAppError("SendNotifications", "app.channel.autofollow.app_error", nil, nErr.Error(), http.StatusInternalServerError)
+					return
+				}
+			}
+			umc <- nil
+		}(id)
+		updateMentionChans = append(updateMentionChans, umc)
```
Comment (jespino) [inline app/notification.go:177]: This is going to create log warnings that was unable to update the channel mention count, that is not related to this. I think we should have another list channels here, to make it independent and generate the logs in the right way.

**Review by jespino — CHANGES_REQUESTED** [review]
Another tiny detail.

File: app/notification.go (hunk 7 above)
Comment (reflog) [reply app/notification.go]: @jespino done

File: app/notification.go [hunk 8]
```diff
@@ -174,7 +175,7 @@ func (a *App) SendNotifications(post *model.Post, team *model.Team, channel *mod
 			}
 			umc <- nil
 		}(id)
-		updateMentionChans = append(updateMentionChans, umc)
+		mentionAutofollowChans = append(mentionAutofollowChans, umc)
```
Comment (jespino) [inline app/notification.go:178]: Probably the channel variable shoud be `mac` here too, instead of reuse the `umc` name.

**Review by jespino — APPROVED** [review]
A cosmetic detail, but otherwise looks good to me.

File: model/config.go [hunk 9]
```diff
@@ -876,6 +876,7 @@ type ExperimentalSettings struct {
 	CloudUserLimit                  *int64  `access:"experimental,write_restrictable,cloud_restrictable"`
 	CloudBilling                    *bool   `access:"experimental,write_restrictable,cloud_restrictable"`
 	EnableSharedChannels            *bool   `access:"experimental"`
+	ThreadAutoFollow                *bool   `access:"experimental"`
```
Comment (esethna) [inline model/config.go:879]: @reflog discussing with Jesus, can we please add this to the service settings section of the config?

**Review by esethna — CHANGES_REQUESTED** [review]

File: model/config.go (hunk 9 above)
Comment (reflog) [reply model/config.go]: @esethna  done

**Comment by reflog** [comment]
@ogi-m - there's not much to review here, no user facing changes. A general smoke test around creating posts with replies

**Review by ogi-m — APPROVED** [review]
Smoke tested posts and post replies, LGTM 👍

**Review by esethna — APPROVED** [review]

**Comment by amyblais** [comment]
@esethna Is this for v5.29?

**Comment by esethna** [comment]
@amyblais yes

**Comment by esethna** [comment]
@amyblais noting that this PR should have a note in the compatibility changelog section, config changelog section and important upgrade notes.

Reason it's needed in the compatibility changelog section is to warn admins that it will be turned on by-default and it may have an impact on performance for large scale organizations with a lot of users.

File: store/sqlstore/thread_store.go [hunk 10]
```diff
@@ -49,6 +54,11 @@ func threadToSlice(thread *model.Thread) []interface{} {
 func (s *SqlThreadStore) createIndexesIfNotExists() {
 	s.CreateIndexIfNotExists("idx_threads_last_reply_at", "Threads", "LastReplyAt")
 	s.CreateIndexIfNotExists("idx_threads_post_id", "Threads", "PostId")
+
+	s.CreateIndexIfNotExists("idx_thread_memberships_last_update_at", "ThreadMemberships", "LastUpdated")
+	s.CreateIndexIfNotExists("idx_thread_memberships_last_view_at", "ThreadMemberships", "LastViewed")
+	s.CreateIndexIfNotExists("idx_thread_memberships_post_id", "ThreadMemberships", "PostId")
```
Comment (mkraft) [inline store/sqlstore/thread_store.go:60]: The `idx_thread_memberships_post_id` index is redundant because the primary key automatically creates a multi-column index for which `PostId` is the first column. The DBs should use that index for `PostId` lookups, so creating this index uses unnecessary storage and slows backups, restores, and writes.

Comment (reflog) [reply store/sqlstore/thread_store.go:60]: @mkraft what about UserId?

Comment (mkraft) [reply store/sqlstore/thread_store.go:60]: I think that's still necessary because it's not the first position of the primary key index. But I would verify it.

Comment (mkraft) [reply store/sqlstore/thread_store.go:60]: This shows that the query plan is almost identical with the difference just in the index used:

```
mattermost_test=# explain (format json) select * from ThreadMemberships where PostId = 'foo';
                        QUERY PLAN
-----------------------------------------------------------
 [                                                        +
   {                                                      +
     "Plan": {                                            +
       "Node Type": "Bitmap Heap Scan",                   +
       "Relation Name": "threadmemberships",              +
       "Alias": "threadmemberships",                      +
       "Startup Cost": 4.16,                              +
       "Total Cost": 9.50,                                +
       "Plan Rows": 2,                                    +
       "Plan Width": 157,                                 +
       "Recheck Cond": "((postid)::text = 'foo'::text)",  +
       "Plans": [                                         +
         {                                                +
           "Node Type": "Bitmap Index Scan",              +
           "Parent Relationship": "Outer",                +
           "Index Name": "idx_thread_memberships_post_id",+
           "Startup Cost": 0.00,                          +
           "Total Cost": 4.16,                            +
           "Plan Rows": 2,                                +
           "Plan Width": 0,                               +
           "Index Cond": "((postid)::text = 'foo'::text)" +
         }                                                +
       ]                                                  +
     }                                                    +
   }                                                      +
 ]
(1 row)
```
```
mattermost_test=# drop index idx_thread_memberships_post_id;
DROP INDEX
```
```
mattermost_test=# explain (format json) select * from ThreadMemberships where PostId = 'foo';
                        QUERY PLAN
----------------------------------------------------------
 [                                                       +
   {                                                     +
     "Plan": {                                           +
       "Node Type": "Bitmap Heap Scan",                  +
       "Relation Name": "threadmemberships",             +
       "Alias": "threadmemberships",                     +
       "Startup Cost": 4.16,                             +
       "Total Cost": 9.50,                               +
       "Plan Rows": 2,                                   +
       "Plan Width": 157,                                +
       "Recheck Cond": "((postid)::text = 'foo'::text)", +
       "Plans": [                                        +
         {                                               +
           "Node Type": "Bitmap Index Scan",             +
           "Parent Relationship": "Outer",               +
           "Index Name": "threadmemberships_pkey",       +
           "Startup Cost": 0.00,                         +
           "Total Cost": 4.16,                           +
           "Plan Rows": 2,                               +
           "Plan Width": 0,                              +
           "Index Cond": "((postid)::text = 'foo'::text)"+
         }                                               +
       ]                                                 +
     }                                                   +
   }                                                     +
 ]
(1 row)
```

Comment (mkraft) [reply store/sqlstore/thread_store.go:60]: And here's the same analysis done in MySQL:
```
mysql> explain format=json select * from ThreadMemberships where PostId = 'foo';
*************************** 1. row ***************************
EXPLAIN: {
  "query_block": {
    "select_id": 1,
    "table": {
      "table_name": "ThreadMemberships",
      "access_type": "ref",
      "possible_keys": [
        "PRIMARY",
        "idx_thread_memberships_post_id"
      ],
      "key": "PRIMARY",
      "used_key_parts": [
        "PostId"
      ],
      "key_length": "106",
      "ref": [
        "const"
      ],
      "rows": 1,
      "filtered": 100,
      "attached_condition": "(`mattermost_test`.`ThreadMemberships`.`PostId` = 'foo')"
    }
  }
}
1 row in set, 1 warning (0.00 sec)
```
```
mysql> drop index idx_thread_memberships_post_id on ThreadMemberships;
Query OK, 0 rows affected (0.01 sec)
Records: 0  Duplicates: 0  Warnings: 0
```
```
mysql> explain format=json select * from ThreadMemberships where PostId = 'foo';
*************************** 1. row ***************************
EXPLAIN: {
  "query_block": {
    "select_id": 1,
    "table": {
      "table_name": "ThreadMemberships",
      "access_type": "ref",
      "possible_keys": [
        "PRIMARY"
      ],
      "key": "PRIMARY",
      "used_key_parts": [
        "PostId"
      ],
      "key_length": "106",
      "ref": [
        "const"
      ],
      "rows": 1,
      "filtered": 100,
      "attached_condition": "(`mattermost_test`.`ThreadMemberships`.`PostId` = 'foo')"
    }
  }
}
1 row in set, 1 warning (0.00 sec)
```

Comment (reflog) [reply store/sqlstore/thread_store.go:60]: thanks for the details, @mkraft . I'm going to remove the first one and leave `idx_thread_memberships_user_id` for now.

**Comment by mkraft** [comment]
@reflog Sounds good. And you probably saw my comment that `idx_threads_post_id` can be removed from the `Threads` table too?

**Comment by reflog** [comment]
@mkraft yup. already done in PR: https://github.com/mattermost/mattermost-server/pull/15994

**Comment by amyblais** [comment]
> @amyblais noting that this PR should have a note in the compatibility changelog section, config changelog section and important upgrade notes.
> 
> Reason it's needed in the compatibility changelog section is to warn admins that it will be turned on by-default and it may have an impact on performance for large scale organizations with a lot of users.

@esethna Changelog and upgrade notes done. Does this need docs in https://docs.mattermost.com/administration/config-settings.html yet?

**Comment by esethna** [comment]
@amyblais yes it should

**Comment by amyblais** [comment]
@esethna Please help submit docs for this for https://docs.mattermost.com/administration/config-settings.html.
